src/types/engine_time.rs

```diff
@@ -0,0 +1,184 @@
+use std::time::Duration;
+
+use crate::raw::root::RED4ext as red;
+
+#[derive(Default, Clone, Copy)]
+#[repr(transparent)]
+pub struct EngineTime(red::EngineTime);
+
+impl EngineTime {
+    pub fn is_valid(&self) -> bool {
+        self.0.unk00 != [0; 8]
+    }
+
+    pub fn as_secs_f64(&self) -> f64 {
+        f64::from_ne_bytes(self.0.unk00)
+    }
+
+    /// SAFETY: panics if the sum ends up being `f64::NAN`, `f64::INFINITY` or `f64::NEG_INFINITY`.
+    pub fn saturating_add_assign(&mut self, value: impl Into<f64>) {
+        let current = self.as_secs_f64();
+        let value: f64 = value.into();
+        let addition = current + value;
+        if addition.is_infinite() {
+            if addition.is_sign_positive() {
+                panic!("EngineTime cannot be infinity");
+            } else {
+                panic!("EngineTime cannot be negative infinity");
+            }
+        } else if addition.is_nan() {
+            panic!("EngineTime cannot be NaN");
+        }
+        self.0.unk00 = addition.to_ne_bytes();
+    }
+
+    pub fn saturating_add(self, value: impl Into<f64>) -> Self {
+        let mut copy = self;
+        copy.saturating_add_assign(value.into());
+        copy
+    }
+
+    /// SAFETY: panics if the sum ends up being `f64::NAN`, `f64::INFINITY` or `f64::NEG_INFINITY`.
+    pub fn saturating_sub_assign(&mut self, value: impl Into<f64>) {
```

Comment:
this is no longer saturating
```suggestion
    pub fn sub_assign(&mut self, value: impl Into<f64>) {
```

Reply:
refactored in 61ec625